doc/man/tones.1
.TH tones 1 "20 Feb 1998" "Release 2.3" "Linux System Manual"
.SH NAME
.I tones \- a sequential tone generator program
.SH SYNOPSIS
.B tones [options] [waveform] T freq(s)|notes(s)|command_file(s)
.SH DESCRIPTION
.I tones
generates one or more tones of various types (waveforms) and duration (T
millisecs) of the specified frequencies or notes, or mixtures of frequencies
or notes. See 
.I tones\ -h
for a list of
possible waveforms. The waveforms should include
.I sine,
.I cosine
(90 degrees out of phase to sine),
.I square
(50% mark/space ratio),
.I sawtooth
(a ramp waveform),
.I triangle
and
.I noise.
Sine is the default. Besides the inbuilt waveforms, waveforms can be loaded
from suitable WAV files - see below
.I LOADABLE\ WAVEFORMS.
.PP
.I T
is the default number of
millisecs that each tone is to be played. 
Frequencies (freq(s)) are specified
in Hertz as integers. A frequency of 0 causes T millisecs of silence to be
played. Notes are specified as the musical note letter with an
optional '#' to sharpen the note, then an octave number. Octaves run from 
C to C. Middle C is C3, the immediately preceding note is B2! The first
Octave is from C0 to C1.
.PP
Several frequencies/notes can be played at once, by specifying the 
frequencies/notes required joined by a ',' character (but no spaces!).
.PP
e.g.
.I 1000,1500,2000 
specifies that the three frequencies are played together, all at the same
relative level. See 
.I AMPLITUDES
sections below for a discussion of how to set absolute amplitude levels
or differing relative amplitudes for notes played together or serially.
.PP
Each freq specification can optionally contain a duration, by
appending ':T', where 'T' is the duration in millisecs. This duration
overrides the default. Also the default duration can be changed 
by using a the ':T' format on it's own - not appended to a freq spec.
.PP
e.g. 
.I 1200,600:1000
play the two freqs for 1 sec
.PP
e.g.
.I c3,e3,g3
play the C major chord
.PP
e.g. 
.I :250
set the default tone duration to 250ms
.PP
Waveforms can be specified/altered at anytime. A single waveform name
specifies that waveform to be used for all channels. Alternatively
a comma (',') seperated list of waveforms can be given to specify or alter
the waveform to use for a given channel. Ommitting a waveform in a list,
means that the previous waveform is left unchanged.
.PP
e.g. 
.I square,,triangle
specifies using square waves for chan 1,
chan 2 is left unchanged, and triangular waves are used for channel 3.
.PP
The digital samples (either 8 or 16 bits) are played by default to the Linux
.I /dev/dsp
device at a samplerate of 22050 samples per second, in mono mode. (see
.I CONFIGURATION\ FILES
section below)
.PP
Fractional Hertz frequencies are not supported. Of course, only frequencies 
less than half the samplerate (number of samples/sec) can be accurately 
generated; but the program doesn't check this.
.PP
Instead of playing the output to 
.I /dev/dsp
the samples can be written to a file as raw samples (-o file) or written in
WAV format (-w wavfile). These data files can then be played back quickly
with a raw data or WAV file player (e.g. wavplay) without the overhead of
actually generating the samples.
.PP
There are some special 'commands' that can be specified, that may be useful
in input files.
.IP \:N
Set default tone duration to N millisecs
.IP \@N
Set base amplitude level of tones when in absolute 
.I amplitude 
mode
.IP absolute
Set absolute amplitude mode (see below)
.IP echo
The rest of the line of the input file, or the rest of the command line
parameter (NB to use quotes where necessary) is output to stdout.
.IP relative
Set relative amplitude mode (see below)
.IP reset|resync
All generator points are reset to the start of the waveform buffers. This
forces subsequant generation of multiple frequencies/waveforms to be in phase.
.PP
Further, if the word is not one of the above, then tones checks to see
if a file of that name exists, and if it does then the file is assumed to be 
a file of tones commands which are executed.
.PP
e.g. 
.I tones\ -v\ :100\ tune1\ tune2\ \ \ \ 
will interpret and play the tones commands in files 
.I tune1
and 
.I tune2.
This file processing is recursive. Files of commands can execute other
files of commands etc. As usual, 
.I '-'
can be used to specify stdin.
.SH RELATIVE\ AMPLITUDES
.PP
.I tones
by default works in a 'relative' amplitude mode, where the output level
and sample range are maximally maintained. This ensures the best signal
accuracy. 
.PP
When specifying multiple frequencies/notes to be played together, then the
relative amplitudes can be specified in deciBells by appending "@db" to the
note.
.PP
e.g.
.I 440,880@-12,1760@-30
specifies a mixture with 880Hz -12dB down, and 1760Hz -30dB down relative
to the level of 440Hz. The mixed signal samples will span the full 16 or 8
bit range permitted for maximal signal accuracy.
.PP
The dB levels indicate the relative power levels.
.I -3dB
being at a relative power level of 0.5,
.I -20dB
being at a relative power level of 0.01 . However power levels are
proportional to the square of the signal amplitude. So a signal at -6dB
(quarter power) will only have its amplitude down by half. To reduce a
signal amplitude by 1/10 then specify -20dB, i.e. a power level down 
by a factor of a hundredth.
.PP 
dB levels can be specified as decimal values.
.SH ABSOLUTE\ AMPLITUDES
.PP
.I tones
can work in an absolute amplitude mode, where 
signal power levels are specified in deciBells (dB) relative to
a 0dB level that indicates a peak value of +32767/-32768 for 16 bit signed
samples,
and 255/0 for 8 bit unsigned values. Hence any signal at a positive dB
level will be clipped. Signals at a negative dB level will attentuated.
If no level is specified then 0dB is assumed.
.PP
e.g.
.I 500@-20,750@-6,1000,-12
gives 500Hz at -20db (amplitude 0.1), 750Hz at -6db (amplitude 0.5),
and 1000Hz at -12dB (amplitude 0.25). The final mixed signal will have
an amplitude of 0.1 + 0.5 + 0.25 = 0.85 or -1.4dB.
.PP
As can be seen, there is no "hands-free" in 
.I absolute
mode. You have to work out the dB levels yourself and ensure that the
resultant mixed signal does not go above 0dB and get clipped. Remember also
that a sine wave at -80dB down (amplitude 1/10000th) only has 6 digital
levels and is a pretty poor representation of a sine wave, not suitable for
post amplification and use!
.PP
In
.I absolute
mode the base 'zero' level can be altered at any time by use of the 
.I @dB
command. All subsequent dB levels specified will have this base level 
added to them.
.PP
e.g.
.I @-20\ 1000,1200@+6,1400@-6
is the same as 1000Hz at -20dB, 1200Hz at -14dB and 1400Hz at -26dB.
.SH LOADABLE\ WAVEFORMS
.PP
Given that the generation method used by 
.I tones
to generate a waveform of FHz is simply to sequentially select 
every Fth sample from a buffer containing S samples of one complete
waveform at a frequency of 1Hz (treating the buffer as circular, the 
beginning conceptually joined to the end), where S is the number of samples
per second, it is possible to load a customised waveform from a WAV file
containing the S samples of a 1Hz waveform. See the
.I -load\ WavFile
and
.I -lw\ N
options below. The name of the waveform is taken as the basename of the
WavFile, i.e. with any trailing '.suffix' and leading path removed.
Each loaded waveform should hence have this name unique, and different from
the inbuilt waveform names.
.PP
The samples in WavFile should be 16 bit, mono, of the same number of samples
as tones' playing samplerate, e.g. if tones is playing at 32000 samples per
sec then the WavFile should contain 32000 16 bit samples. 16 bit samples are
needed, because tones works internally with 16 bit samples, even if it is
feeding 8 bit samples to the sound card or output file. Ideally the samples
should span one complete wavelength, i.e. represent 1 second of a 1Hz signal.
However this can be varied if used with some intelligence. If, say, 1 seconds
worth of 5Hz of the waveform is used, then the output frequency will be 5
times higher than specified. If you have a mixture of 3Hz and 5Hz samples,
then the frequencies generated will be a mixture of 3 and 5 times the
frequency specified. I hope that is all understandable!
.PP
See the 
.I tones.eg
directory for some examples of loadable modules and how
.I tones
itself can be used to generate the loadable waveforms.
.SH OPTIONS
.IP -8\ |\ -b\ 8
set 8 bit unsigned data samples
.IP -16\ |\ -b\ 16
set 16 bit signed little-endian data samples.
.IP -abs|-absolute
set absolute amplitude mode
.IP -a
when used in conjunction with the -o option, data is appended to the file.
.IP -C\ file
use "file" as the local configuration file (see below).
.IP -c\ CHANNELS
set the maximum number of channels (concurrent played frequencies) to 
CHANNELS. The default number is 4. There is some virtue in keeping the
number of channels to a minimum.
.IP -f
when used in conjunction with the -o or -w options, any existing file is
silently overwritten.
.IP -h
display usage and help info
.IP -i\ file
read frequencies/waveforms to generate from file 'file'.
Reads from standard input if filename is '-'. Any command line specifications
are actioned before the input file is read.
.IP -l
play the tone sequence repetitively. Forced off if writing samples to a file
with the -o or -w options.
.IP -loop\ N
play the tone sequence N times.
.IP -o\ file
write out samples to a raw data file. You will have to remember the data
format, e.g. samplerate and 8/16 bit.
.IP -rel|-relative
set relative amplitude mode
.IP -s\ samplerate
set the number of samples per second to samplerate. For many simple uses a
samplerate of 8000 is sufficient, making any saved data files smaller.
.IP -w\ wavfile
write samples out in WAV format to wavfile. The WAV header contains details
of whether the data is 8 or 16 bits and the sampling rate. You cannot use the
append (-a) option with WAV files.
.IP -v
be verbose
.IP -lw\ N
Specify the number of loadable waveforms allowed, the default is 4
.IP -load\ WavFile
Load the waveform from the WavFile.
.SH EXAMPLES
.IP tones\ 50\ 1000\ 700,1200\ 800,1100,1300
generates 3 50 millisecs sine tones, the first consisting of only 1000Hz,
the second of 700Hz and 1200Hz and the third of 800Hz, 1100Hz and 1300Hz
.IP tones\ -loop\ square\ 200\ 700\ 900\ 400\ 500
generates a sequence of 4 200 millisecs square wave tones which is repeated
until the program is interupted.
.IP tones\ -w\ seq.wav\ 70\ 1016\ 1200\ 1080\ 1150\ 1016
generates a sequence of 5 70 millisecs sine tones, and instead of playing them
the samples are stored in WAV format in seq.wav which can be played by any
WAV file player.
.IP tones\ -w\ trap.wav\ :1000\ triangle\ absolute\ 1@6
Generates a WAV file
.I trap.wav
consisting of a trapezoid waveform where the rise and fall slopes take up
half the wavelength. A sawtooth is generated with a maximum that has twice
the amplitude of the maximum sample sizes allowed, hence it is clipped flat
for half the waveform period making a trapezoid shape.
.IP tones\ -load\ trap.wav\ :1000\ triangle\ 1000\ trap\ 500\ triangle\,trap\ 1000\,500
Will load the trapezoid waveform generated above as a new waveform called
.I trap
and then plays 1 seconds each of first a 1000Hz triangle wave, then a 500Hz
.I trap
waveform and finally both waveforms played together.
.PP 
See also the 
.I tones.eg 
directory in the siggen distribution.
.SH CONFIGURATION\ FILES
.PP
Three possible configuration files can be used: a LOCAL config file (usually
in current directory), a HOME config file in user's $HOME directory and a
GLOBAL config file.
.PP
All the siggen suite of programs are compiled with the names of the config
files built in. By default the configuration files are:
.IP ./.siggen.conf
is the LOCAL config file.
.IP $HOME/.siggen.conf
is the HOME config file.
.IP /etc/siggen.conf
is the GLOBAL config file.
.IP tones\ -h
will indicate which config files will be searched for.
.PP
The config files do not have to exist. If they exist and are readable by the
program they are used, otherwise they are simply ignored.
.PP
The config files are always searched for configuration values in the order
LOCAL, HOME, GLOBAL. This allows a scheme where the sysadmin sets up default
config values in the GLOBAL config file, but allows a user to set some or
all different values in their own HOME config file, and to set yet more
specific values when run from a particular directory.
.PP
If no configuration files exist, the program provides builtin
default values, and most of these values can be set
by appropriate command line switches and flags.
.PP
See siggen.conf(5) for details of the configuration files.
.PP
.I tones
looks for configuration values CHANNELS, DACFILE, SAMPLERATE,
SAMPLESIZE, VERBOSE, LOADABLE_WAVEFORMS.
.IP CHANNELS
sets the number of channels, see '-c' option.
.IP DACFILE
allows the name of the DAC/DSP/PCM device to be changed from /dev/dsp
.IP LOADABLE_WAVEFORMS
specifies the allowable number of loadable waveforms
.IP SAMPLERATE
sets the number of samples/sec for the DAC device
.IP SAMPLESIZE
sets whether 8 or 16 bit samples to be generated
.IP VERBOSE
sets whether or not to run in verbose mode.

.SH SEE  ALSO
.IP siggen.conf(5),\ sgen(1),\ swgen(1)
.SH
.SH BUGS
.SH
.SH COPYING
.I Copyright\ 1995-2008\ Jim\ Jackson
.PP
The software described by this manual is covered by the GNU General
Public License, Version 2, June 1991, issued by :
.IP
Free Software Foundation, Inc.,
.br
675 Mass Ave,
.br
Cambridge, MA 02139, USA
.PP
Permission is granted to make and distribute verbatim copies of
this manual provided the copyright notice and this permission notice
are preserved on all copies.
.PP
Permission is granted to copy and distribute modified versions of this
manual under the conditions for verbatim copying, provided that the
entire resulting derived work is distributed under the terms of a
permission notice identical to this one.
.PP
Permission is granted to copy and distribute translations of this
manual into another language, under the above conditions for modified
versions, except that this permission notice may be included in
translation instead of in the original English.
.SH AUTHOR
.I Jim Jackson
.br
.sp
.I Email: jj@franjam.org.uk
.br


CHANGES

siggen - a package of waveform generation programs for linux
------------------------------------------------------------

Version 2.3.+

  siggen
  
    Tried adding loadable waveforms - came to an impasse. I will revisit
    this.
    Added a "tracking" feature suggested by Gart Reese. 't' or 'T'
    while in the waveform, frequency or gain fields will cause both
    similar fields of each channel to track, i.e. have the same value.
    A "tracking" message shows which fields are currently tracking.
    Repeating 't' or 'T' cancels the tracking of the fields.
    
  fsynth
    
    A curses based "fourier" synthesis program. Allows a mixture of
    a fundamental sine wave and varying proportions of it's harmonics
    to be mixed to create complex waveforms.
    
  tones

    Added loadable waveforms. Tones can load a waveform from a WAV file
    that contains exactly one seconds worth of samples, at the same samplerate
    as tones is using and same format etc, and it will "play" from the these
    samples. The waveform name is the basename of the wav file with directory
    and any suffix removed. To work correctly the samples should be for a 1Hz
    wave. If its for say a 2Hz wave then the frequencies generated will all
    be double for that waveform.
    Added recursive command file reading.
    Attenuation/gain levels added, thanks to "Mark E. Shoulson"
    <mark@kli.org> for a patch, feedback and comments. Fixed faulty mixing
    code for more than 2 channels.

  All curses progs

    The ability to "Lock" input field to freeze values. Locked fields
    are skipped while moving around the screen. All locked fields can 
    be unlocked in one go to reset. '!' or 'L' locks the current field,
    'U' unlocks all locked fields. This was the result of misunderstanding
    an idea from Garst!
    
    Also added ability to redraw screen by pressing Ctrl-L (^L), useful
    if screen gets garbled.

    Bug fix in scfio.c, thanks to "Garst R. Reese" <reese@isn.net>
    for use egcs and finding the bug - I think gcc was too forgiving.

  All progs
    
    Added configuration file(s) support. 

Version 2.2

  tones, sgen, swgen

    Added ability to compile on machines without sound support,
    so that WAV and raw data files can be created.

    Fixed so can be compiled under SUNOS on sparc machines. This meant
    fixing the implicit little endian nature of the raw and wav file writing
    code.
   
    New config.h file added with user configurable options gathered together
    in one place, for global options.

  tones, sgen, swgen

    Added the -x10 and -x100 options, which when in use allow the
    generation of freq. to the nearest 0.1Hz or 0.01Hz by forcing
    the specification of all frequencies to be 10 or 100 times
    too big. 

    e.g.  swgen -x10 -2 -v 2 2000 20000

          causes stereo samples for a 0.2Hz sawtooth ramp sweeping
          a sin wave from 200Hz to 2000Hz.

Version 2.1

   Fixed compilation problem of incompatible extern declaration
     of sys_errlist . extern wasn't necessary so deleted it.

Version 2

   Added swgen - a swept frequency generator.

   Added sweepgen - a curses based swept freq. generator

   Changed siggen to play continuously - changes to waveform, freq,
     gain etc take effect (almost) immediately. Downside is that
     this is a lot more processor intensive. Fine, just, on my 486DX2/66.

   Extended the tones program, extensively.

   Added dtmf program which can be used as tone phone dialer.

   Fixed mixed up description of sawtooth/triangle waves in sgen manual.

Version 1.1  released xx March 1997

   Fixed a small initialise bug in sgen - reported by 
      Michael Meifert <mime@dk3hg.hanse.de>  
      (Michael this email address doesn't work)

   Fixed a bug in siggen that prevented PULSE waveforms being generated
      
   Added the tones program following a suggestion by Michael who
      wanted a program to generate tones to control a radio repeater (I think).


Version 1    released  6 March 1997

   contained programs sgen, siggen, smix, soundinfo




src/tones.c
/* tones.c
 * generate and play tones for given period....
 * Jim Jackson     Linux Version
 */

/*
 * Copyright (C) 1997-2008 Jim Jackson               jj@franjam.org.uk
 * 
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program - see the file COPYING; if not, write to 
 *  the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, 
 *  MA 02139, USA.
 */

/* History...................
 * 
 * 17Mar06  Fixed Wav writing bug reported by Lukas Loehrer
 *  9Mar06  Fixed raw writing bug reported by Lukas Loehrer
 *   Apr00  added loadable waveforms
 * 18May99  added absolute level mode.
 *  1Sep98  added configuration file(s) support
 * 18Jun98  Modified to compile with no sound card support etc and under SUNOS
 * 20Feb98  Added recognition of musical notes format e.g. G3 for G
 *          in 3rd octave, A2# for A# in second octave. Octave '0'
 *          runs from C at 33Hz to C at 65Hz
 *  9Sep97  Split out the tone spec. parsing and tone playing into 
 *          tonesgen.c
 *  8Sep97  Alter tones specification to be able to specify different
 *          durations. i.e. t1,t2:ms. 
 *          Also start the change of generating technique to do sampling
 *          1Hz sample space for requisite number of samples.
 * --Aug97  Re-organise some of DAC routines out into DAC.c
 * 18Feb97  -f flag missing, added it
 *          also added facility to be able to change wavform. A
 *          non-numeric parameter will cause a change in waveform
 *          for rest of tones.
 *          Also still occasional noise after tones played - altered DSPend again!
 * 16Feb97  altered it so you can specify several freqs to be played together
 *          by using punctuation of some sort to seperate freqs 
 *          e.g. 440,880   would play two freqs of 440Hz and 880Hz together
 * 13Feb97  Based on sgen, using generator.c - only mono version.
 * 
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <ctype.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <math.h>
#include "config.h"

#define VERSTR "%s  Ver. %s   Tone Generator.\n"

#define EFAIL -1

#define MAXPRM 32
#define MAXTONES 64
#define chkword(a,b) ((n>=a)&&(strncmp(p,b,n)==0))

int aflg,fflg,vflg,dflg,wfmt;
int loopflg;

int Abs_Level ;                  /* if set cancel auto-scale mode */
unsigned int stereo;             /* stereo or mono mode      */
unsigned int samplerate;         /* Samples/sec              */
unsigned int ratio;              /* used in pulse, sweep etc */
unsigned int afmt;               /* format for DSP  */
unsigned int DSPbytecount;       /* number of bytes sent to DSP  */
int frag_size;                   /* size of a DSP buffer  */
int channels;                    /* number of generating channels */
int Wn;                          /* number of allowed loadable waveforms */
int TWn;                         /* Total number of waveforms  */
int LWn;                         /* number of specified loadable waveforms */
char **LWaa;                     /* array of specifed loadable waveforms */

char *sys,*outfile,*infile;
char dac[130];
char *configfile;

extern char *getGenErrStr();

help(e)
int e;
{  
   char **aa;
   
   fprintf(stderr,VERSTR,sys,VERSION);
   fputs("\nUsage: tones [-s samplerate] [-8/-16|-b 8/16] [waveform] :T freq(s)\n\n",stderr);

#ifdef HAVE_DAC_DEVICE
   fprintf(stderr,"Defaults: SINE wave, %d channels, output to %s, %d samples/sec,\n",
	   DEF_TONES_CHANNELS,DAC_FILE,SAMPLERATE);
   fputs("         16 bit mono samples if possible, else 8 bit.\n",stderr);
#else
   fprintf(stderr,"Defaults: SINE wave, %d channels, %d samples/sec,\n",
	   DEF_TONES_CHANNELS,SAMPLERATE);
   fputs("         16 bit mono samples. Must be used with -o or -w option.\n",stderr);
#endif
   fputs("\
Generates T millisecs of each frequency specified. Several freqs can be\n\
played together, e.g. 150,200,300. Relative amplitudes can be specified\n\
in dB e.g. 150@-6,200,300@-12. The channels option gives the maximum number\n\
of freqs. Each channel adds processing overhead to the generation process.\n\
Non-default specific durations maybe given e.g. 440,880:1000\n\
",stderr);
#if FALSE
   fputs("Generates tones of given frequencies for T millisecs each.\n",stderr);
   fputs("Several freqs can be played together by specifying a list of freqs\n",stderr);
   fputs("separated by ',', e.g. 150,200,300. relative amplitudes can be specified\n",stderr);
   fputs("by using the suffix @-dB after the freq.\n",stderr);
   fputs("The maximum number of frequencies which can be played together\n",stderr);
   fputs("is specified by the '-c channels' option. Each channel adds\n",stderr);
   fputs("processing overhead to the generation process.\n",stderr);
   fputs("A specific duration of N millisecs maybe given, overriding the default,\n",stderr);
   fputs("by appending ':N'\n",stderr);
   fputs("e.g. 150,300:250 plays 150Hz and 300Hz together for 250 millisecs\n",stderr);
   fputs("     0:1000 plays one second (1000ms) of silence.\n",stderr);
#endif
   fputs("Valid waveforms are:",stderr);
   for ( aa=(char **)getWavNames(); *aa; aa++ ) fprintf(stderr," %s",*aa);
   fputs(".\nand can be specified anywhere and affect all later tones.\n",stderr);
   fputs("flags: -f,-a         force overwrite/append of/to file\n",stderr);
   fputs("       -o/-w file    write samples to raw/WAVE file ('-' is stderr)\n",stderr);
   fputs("       -i file       contents of file specify tones program ('-' is stdin)\n",stderr);
   fputs("       -load wave_file  load waveform from wav_file\n",stderr);
   fputs("       -s samples    generate with samplerate of samples/sec\n",stderr);
   fputs("       -c channels   number of channels or voices\n",stderr);
   fprintf(stderr,"       -lw N         number of loadable waveforms allowed - def %d\n",DEF_WN);
   fputs("       -loop N       repeat play the tones N times.\n",stderr);
   fputs("       -l            repeat play the tones ad nauseam.\n",stderr);
   fputs("      -v/-vv         be verbose / be very verbose\n",stderr);
   fputs(" -8/-16 or -b 8|16   force 8 bit or 16 bit mode.\n",stderr);
   fputs("    -abs|-rel        set absolute or relative amplitude control.\n",stderr);
   return(e);
}

/* main
 *
 */
 
main(argc,argv)
int argc;
char **argv;
{
   unsigned int i,j,k,l,m,n,N;
   FILE *f,*fi;
   char *p,*fnm,bf[260],*omode;
   int fd,st;
   unsigned int t;
   
   if ((p=strrchr(sys=*argv++,'/'))!=NULL) sys=++p;
   argc--;
   
   configfile=DEF_CONF_FILENAME;
   outfile=infile=NULL;
   samplerate=0; afmt=AFMT_QUERY;
   channels=0; Wn=-1; LWn=0; LWaa=argv;
   stereo=0;
   loopflg=wfmt=frag_size=0;                 /* force mono mode */
   Abs_Level=dflg=vflg=aflg=fflg=0;
   
   while (argc && **argv=='-') {          /* all flags and options must come */
      n=strlen(p=(*argv++)+1); argc--;    /* before paramters                */
      if (chkword(1,"samplerate")) {
	 if (argc && isdigit(**argv)) { samplerate=atoi(*argv++); argc--; }
      }
      else if (chkword(1,"output")) {     /* specify output file       */
	 if (argc) { outfile=*argv++; argc--; } /* output file name          */
      }
      else if (chkword(1,"input")) {     /* specify input file       */
	 if (argc) { infile=*argv++; argc--; } /* input file name          */
      }
      else if (chkword(3,"load")) {     /* load a waveform           */
	 if (argc && **argv!='-') { LWaa[LWn++]=*argv++; argc--; } /* waveform file name   */
      }
      else if (chkword(2,"lw")) {
	 i=0;
	 if (argc && **argv!='-') { 
	    i=atoi(*argv++); argc--;
	    Wn=(i>0)?i:0;
	 }
      } 
      else if (chkword(3,"loop")) {       /*  specify loop option   */
	 if (argc && **argv!='-') {
	    loopflg=atoi(*argv++); argc--;
	 }
      }
      else if (chkword(3,"absolute")) {   /*  absolute amplitude levels */
         Abs_Level=1;
      }
      else if (chkword(3,"relative")) {   /*  relative amplitude levels */
         Abs_Level=0;
      }
      else if (chkword(1,"wave")) {     /* e.g. write WAVE format file  */
	 if (argc) { outfile=*argv++; argc--; wfmt=1; } /* output file name  */
      }
      else if (chkword(2,"16")) { afmt=AFMT_S16_LE; }
      else if (chkword(1,"bits")) {
	 i=0;
	 if (argc) { 
	    i=atoi(*argv++); argc--;
	 }
	 if (i==8) afmt=AFMT_U8;
	 else if (i==16) afmt=AFMT_S16_LE;
	 else exit(err_rpt(EINVAL,"must be '-b 8' or '-b 16'."));
      }
      else if (chkword(1,"channels")) {
	 i=0;
	 if (argc) { 
	    i=atoi(*argv++); argc--;
	 }
	 if (i<1) exit(err_rpt(EINVAL,"Must have 1 or more channels."));
	 channels=i;
      }
      else {                             /* check for single char. flags    */
	 for (; *p; p++) {
	    if (*p=='h') exit(help(EFAIL));
	    else if (*p=='8') afmt=AFMT_U8;
	    else if (*p=='a') aflg=1;
	    else if (*p=='f') fflg=1;
	    else if (*p=='d') dflg=1;
	    else if (*p=='l') loopflg=-1;
	    else if (*p=='v') vflg++;
	    else {
	       *bf='-'; *(bf+1)=*p; *(bf+2)=0;
	       exit(help(err_rpt(EINVAL,bf)));
	    }
	 }
      }
   }
   
  /* interrogate config file....... */

   init_conf_files(configfile,DEF_CONF_FILENAME,DEF_GLOBAL_CONF_FILE,vflg);
   if (vflg==0) {
      vflg=atoi(get_conf_value(sys,"verbose","0"));
   }
   if (samplerate==0) {
      samplerate=atoi(get_conf_value(sys,"samplerate",QSAMPLERATE));
   }
   if (channels==0) {
      channels=atoi(get_conf_value(sys,"channels",QDEF_TONES_CHANNELS));
   }
   if (Wn==-1) {
      Wn=atoi(get_conf_value(sys,"loadable_waveforms",QDEF_WN));
   }
   if (afmt==AFMT_QUERY) {
      afmt=atoi(get_conf_value(sys,"samplesize",QAFMT_QUERY));
   }
   strncpy(dac,get_conf_value(sys,"dacfile",DAC_FILE),sizeof(dac)-1);

  /* Now check if any input file to be read........ */
  
   if (vflg) printf(VERSTR,sys,VERSION);

   fi=NULL;            /* fi!=NULL if a command file is to be read */
   if (infile!=NULL) { /* check for valid infile if specified */
      if (strcmp(infile,"-")==0) {
	 fi=stdin; infile="stdin";
      } else if ((fi=fopen(infile,"r"))==NULL) {
	 exit(err_rpt(errno,infile));
      }
   }

   /* open the right device or output file............ */
   f=NULL;
   if (outfile==NULL) {
#ifdef HAVE_DAC_DEVICE
                                  /* if no outfile then write direct to DAC */
                                  /* if no format specified then try 16 bit */
      i=(afmt==AFMT_QUERY)?AFMT_S16_LE:afmt ; 
      if ((fd=DACopen(fnm=dac,"w",&samplerate,&i,&stereo))<0) {
	 if (afmt==AFMT_QUERY) {  /* if no format specified try for 8 bit.. */
	    i=AFMT_U8;
	    fd=DACopen(fnm,"w",&samplerate,&i,&stereo);
	 }
	 if (fd<0) exit(err_rpt(errno,fnm));
      }
      if ((frag_size=getfragsize(fd))<0)
	exit(err_rpt(errno,"Problem getting DAC Buffer size."));

      if (vflg) {
	 printf("%s : DAC Opened for output\n",fnm);
         printf("%d bytes per DAC buffer.\n",frag_size);
      }
      if ((afmt!=AFMT_QUERY && i!=afmt) || stereo!=0) {
	 if (i!=afmt) 
	   err_rpt(EINVAL,"Sound card doesn't support requested sample format.");
	 if (stereo!=0)
	   err_rpt(EINVAL,"Card doesn't do MONO - nah, don't believe it.");
	 exit(EINVAL);
      }
      afmt=i;
#else
                                 /* error if no outfile specified */
      exit(help(err_rpt(EINVAL,"No output file specified, use -w or -o option.")));
#endif
   }
   else {
     if (strcmp(outfile,"-")==0) {
       if (aflg)
         exit(err_rpt(EINVAL,"Cannot write append to stdout."));
       f=stdout; fnm="stdout";
     } else {
       if (aflg && wfmt) {
         exit(err_rpt(EINVAL,"Cannot write append to a WAVE file."));
       }
       afmt=(afmt==AFMT_QUERY)?AFMT_S16_LE:afmt ; /* set 16 bit mode unless
						   * some format is forced */
       omode=(aflg)?"a":"w";
       if ((f=fopen(fnm=outfile,"r"))!=NULL) {
	 fclose(f);
	 if ( (!aflg) && (!fflg) ) exit(err_rpt(EEXIST,fnm));
       }
       if ((f=fopen(fnm,omode))==NULL) { exit(err_rpt(errno,fnm)); }
       frag_size=32768;
       if (vflg) { 
	 printf("%s samples to %s file %s\n",
		(*omode=='a')?"Appending":"Writing",
		(wfmt)?"WAVE":"RAW", fnm);	       
         printf("Buffer size is %d bytes.\n",frag_size);
       }
       fd=fileno(f);
       if (loopflg<0) loopflg=0; /* cannot loop forever if writing to a file */
       if (wfmt && (fWAVwrhdr(f,0,samplerate,afmt,stereo)<0)) {
	 exit(err_rpt(errno,"Writing WAVE file header."));
       }
     }
     if (fWAVinit(0,samplerate,afmt,stereo)) {
	 exit(err_rpt(errno,"Initialising sound file parameters."));
     }
   }
   if (!(afmt==AFMT_S16_LE || afmt==AFMT_U8)) {
      exit(err_rpt(EINVAL,
		   "Only unsigned 8 and signed 16 bit samples supported."));
   }
 	
   if (vflg) {
     if (vflg>1) fWAVdebug(stdout);
     printf("%d %s, %s, samples/sec\n",samplerate,
	    (stereo)?"stereo":"mono",
	    (afmt==AFMT_U8)?"unsigned 8 bit":
	    ((afmt==AFMT_S16_LE)?"signed 16 bit, little endian":
	     "Unknown audio format"));
     printf("Upto %d Frequencies can be played together.\n",channels);
     printf("%d loadable waveforms specified.\n",LWn);
     printf("Tone levels are %s.\n",(Abs_Level)?"ABSOLUTE":"RELATIVE");
   }

   /* initialise base samples etc */
   TWn=LWn;
   init_samples(f,fd,frag_size,samplerate,afmt,stereo,channels,&TWn);
   if (LWn && (st=do_loadable_waveforms(LWaa,LWn,TWn,vflg))) exit(st);

   do {
      for ( i=0; i<argc; i++) {
	 if (vflg) printf("%s\n",argv[i]);
	 if (st=doToneCmd(argv[i])) {
	    fprintf(stderr,"%s -> %s\n",argv[i],getGenErrStr());
	 }
      }
      if (fi!=NULL) ftones(fi,infile);
      if (loopflg>0) loopflg--;
   } while (loopflg);
   
   if (outfile==NULL) {
     flush_samples();
     close(fd);
   } else {
     if (wfmt && 
	 (fWAVwrhdr(f,getsamples(),samplerate,afmt,stereo)<0)) {
       err_rpt(errno,"Rewriting WAVE file header.");
     }
     fclose(f);
   }
   
   exit(0);
}  
